pynamodb/connection/base.py

```diff
@@ -899,34 +947,86 @@ def put_item(self,
-
-        operation_kwargs = {TABLE_NAME: table_name}
-        operation_kwargs.update(self.get_identifier_map(table_name, hash_key, range_key, key=ITEM))
-        name_placeholders = {}
-        expression_attribute_values = {}
+        operation_kwargs = self.get_operation_kwargs(
+            table_name=table_name,
+            hash_key=hash_key,
+            range_key=range_key,
+            key=ITEM,
+            attributes=attributes,
+            condition=condition,
+            return_values=return_values,
+            return_consumed_capacity=return_consumed_capacity,
+            return_item_collection_metrics=return_item_collection_metrics
+        )
+        try:
+            return self.dispatch(PUT_ITEM, operation_kwargs)
+        except BOTOCORE_EXCEPTIONS as e:
+            raise PutError("Failed to put item: {}".format(e), e)
 
-        if attributes:
-            attrs = self.get_item_attribute_map(table_name, attributes)
-            operation_kwargs[ITEM].update(attrs[ITEM])
-        if condition is not None:
-            condition_expression = condition.serialize(name_placeholders, expression_attribute_values)
-            operation_kwargs[CONDITION_EXPRESSION] = condition_expression
-        if return_consumed_capacity:
+    def _get_transact_operation_kwargs(self,
+                                       client_request_token=None,
+                                       return_consumed_capacity=None,
+                                       return_item_collection_metrics=None):
+        operation_kwargs = {}
+        if client_request_token is not None:
+            operation_kwargs[CLIENT_REQUEST_TOKEN] = client_request_token
+        if return_consumed_capacity is not None:
             operation_kwargs.update(self.get_consumed_capacity_map(return_consumed_capacity))
-        if return_item_collection_metrics:
+        if return_item_collection_metrics is not None:
             operation_kwargs.update(self.get_item_collection_map(return_item_collection_metrics))
-        if return_values:
-            operation_kwargs.update(self.get_return_values_map(return_values))
-        if name_placeholders:
-            operation_kwargs[EXPRESSION_ATTRIBUTE_NAMES] = self._reverse_dict(name_placeholders)
-        if expression_attribute_values:
-            operation_kwargs[EXPRESSION_ATTRIBUTE_VALUES] = expression_attribute_values
+
+        return operation_kwargs
+
+    def transact_write_items(self,
+                             condition_check_items,
+                             delete_items,
+                             put_items,
+                             update_items,
+                             client_request_token=None,
+                             return_consumed_capacity=None,
+                             return_item_collection_metrics=None):
+        """
+        Performs the TransactWrite operation and returns the result
+        """
+        transact_items = []
+        transact_items.extend([
```

Comment:
nit: the explicit `[]` wrapping should be unnecessary -- extend can take an iterable. i.e

```python
transact_items.extend(
  {...} for foo in bar
)
```